src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -0,0 +1,668 @@
+-- basic associated functions.
+----------------------------------------------------------------------------
+module Algebra.Graph.Bipartite.AdjacencyMap (
+    -- * Data structure
+    AdjacencyMap, leftAdjacencyMap, rightAdjacencyMap,
+
+    -- * Basic graph construction primitives
+    empty, leftVertex, rightVertex, vertex, edge, overlay, connect,
+    vertices, edges, overlays, connects, swap,
+
+    -- * Conversion functions
+    toBipartite, fromBipartite, fromGraph,
+
+    -- * Graph properties
+    isEmpty, hasEdge, hasLeftVertex, hasRightVertex, hasVertex, leftVertexCount,
+    rightVertexCount, vertexCount, edgeCount, leftVertexList, rightVertexList,
+    vertexList, edgeList, leftVertexSet, rightVertexSet, vertexSet, edgeSet,
+
+    -- * Miscellaneous
+    consistent,
+    ) where
+
+import Data.Either (lefts, rights)
+import Data.List   (sort, (\\))
+import GHC.Generics
+
+import qualified Algebra.Graph              as G
+import qualified Algebra.Graph.AdjacencyMap as AM
+
+import qualified Data.Map.Strict as Map
+import qualified Data.Set        as Set
+import qualified Data.Tuple
+
+{-| The 'Bipartite.AdjacencyMap' data type represents an __undirected__
+bipartite graph by two maps of vertices into their adjacency sets. The two type
+parameteters define the types of identifiers of the vertices of each part.
+
+__Note:__ even if the identifiers and their types for two vertices of
+different parts are equal, these vertices are considered to be different.
+See examples for more details.
+
+We define a 'Num' instance as a convenient notation for working with bipartite
+graphs:
+
+@
+0                         == rightVertex 0
+'swap' 1                    == leftVertex 1
+'swap' 1 + 2                == vertices [1] [2]
+('swap' 1) * 2              == edge 1 2
+('swap' 1) + 2 * ('swap' 3)   == overlay (leftVertex 1) (edge 3 2)
+('swap' 1) * (2 + ('swap' 3)) == connect (leftVertex 1) (vertices [3] [2])
+@
+
+__Note:__ the 'Num' instance does not satisfy several "customary laws" of 'Num',
+which dictate that 'fromInteger' @0@ and 'fromInteger' @1@ should act as
+additive and multiplicative identities, and 'negate' as additive inverse.
+Nevertheless, overloading 'fromInteger', '+' and '*' is very convenient when
+working with algebraic graphs; we hope that in future Haskell's Prelude will
+provide a more fine-grained class hierarchy for algebraic structures, which we
+would be able to utilise without violating any laws.
+
+The 'Show' instance is defined using basic graph construction primitives:
+
+@
+show (empty)                       == "empty"
+show 1                             == "rightVertex 1"
+show ('swap' 2)                      == "leftVertex 2"
+show (1 + 2)                       == "vertices [] [1,2]"
+show ('swap' (1 + 2))                == "vertices [1,2] []"
+show (('swap' 1) * 2)                == "edge 1 2"
+show (('swap' 1) * 2 * ('swap' 3))     == "edges [(1,2),(3,2)]"
```

Comment:
Note that functions bind more tightly than operators, so some of the above parentheses are redundant (also in tests).